src/web/js/cpo-ide-hooks.js

```diff
@@ -35,22 +39,86 @@
+      return runtime.safeCall(function() {
+        return runtime.ffi.cases(gmf(compileStructs, "is-Dependency"), "Dependency", dependency,
+          {
+            builtin: function(name) {
+              var raw = cpoModules.getBuiltinLoadableName(runtime, name);
+              if(!raw) {
+                throw runtime.throwMessageException("Unknown module: " + name);
+              }
+              else {
+                return gmf(cpo, "make-builtin-js-locator").app(name, raw);
+              }
+            },
+            dependency: function(protocol, args) {
+              console.error("Unknown import: ", dependency);
+            }
+          });
+       }, function(l) {
+          return gmf(compileLib, "located").app(l, runtime.nothing);
+       }, "findModule");
+    }
+    var pyFindModule = runtime.makeFunction(findModule, "find-module");
+
+    // NOTE(joe): This line is "cheating" by mixing runtime levels, and uses
+    // the same runtime for the compiler and running code.  Usually you can
+    // only get a new pyret-viewable Runtime by calling create, but here we
+    // magic the current runtime into one.
+    // Someday Pyret will be quick enough that we won't need to save theses
+    // seconds of instantiation.
+    var pyRuntime = gf(gf(runtimeLib, "internal").brandRuntime, "brand").app(
+      runtime.makeObject({
+        "runtime": runtime.makeOpaque(runtime)
+      }));
+    var pyRealm = gf(loadLib, "internal").makeRealm(cpoModules.getRealm());
+
+    var builtins = [];
+    Object.keys(runtime.getParam("staticModules")).forEach(function(k) {
+      if(k.indexOf("builtin://") === 0) {
+        builtins.push(runtime.makeObject({
+          uri: k,
+          raw: cpoModules.getBuiltinLoadable(runtime, k)
+        }));
+      }
+    });
+    var builtinsForPyret = runtime.ffi.makeList(builtins);
+
+    function parse(source, uri) {
+      var parse = runtime.getField(runtime.getField(parsePyret, "values"), "surface-parse")
+      return runtime.safeTail(function() {
+        return parse.app(source, uri);
+      });
+    }
+
+    function compile(ast) {
+      var compileAst = gmf(cpo, "compile-ast");
+      return runtime.safeTail(function() {
+        return compileAst.app(ast, pyRuntime, pyFindModule, gmf(compileStructs, "default-compile-options"));
+      });
+    }
+
+    function run(jsSrc) {
+      var run = gmf(cpo, "run");
+      return runtime.safeTail(function() {
+        return run.app(pyRuntime, pyRealm, jsSrc);
+      });
+    }
+
     window.CPOIDEHooks = {
       runtime: runtime,
-      namespace: namespace,
-      uri: uri,
-      compileLib: compileLib,
-      compileStructs: compileStructs,
-      pyRepl: pyRepl,
-      cpo: cpo,
-      runtimeLib: runtimeLib,
-      loadLib: loadLib,
-      builtinModules: builtinModules,
-      cpoBuiltins: cpoBuiltins,
-      cpoModules: cpoModules,
+      parse: parse,
+      compile: compile,
+      run: run
```

Comment:
Is there any reason not to take the parse/compile/run functions you've created here and just move them up into ide.js? 

The main benefit of moving this logic up the stack is that it becomes easier to change without going through the pyret build process, which at least at this particular moment is a bit more cumbersome than webpack :)